[IMP] web: Company service refactor and improved ease of understanding

BastienFafchamps opened the pull request:
This commit refactors the company service to be split responsibilities. The selection logic is now entirely contained in the switch_company_menu, making the company service clearer.

The following variables have also been renamed:
- `availableCompanies` has been renamed to `allowedCompanies` to reflect the name from the python side
- `allowedCompanies` has been renamed to `activeCompanies`

Some features were missing on mobile:
- Indent has been added to SwitchMenuItem to display the hierachy of companies on mobile.
- Unallowed ancestor companies where not displayed on mobile.

*The functional aspect of the company service and company switch has not been changed.*

Enterprise PR: https://github.com/odoo/enterprise/pull/48862
Task: [3522168](https://www.odoo.com/mail/view?model=project.task&res_id=3522168)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/138521).

tde-banana-odoo commented:
I would not rename the variable about allowed companies, as it matches what is used server-side. It always help to keep variables synchronized; otherwise you end up wondering how to find allowed_companies in python, nobody will suddenly guess it became activeCompanies. As I frequently have to guess what is the matching between python and JS, unless you have strong reasons not to do so, please keep the naming for people who have some history in Odoo addons.

Cheers !

aab-odoo commented on addons/web/static/src/webclient/company_service.js:
```diff
@@ -84,43 +64,142 @@ export const companyService = {
+/**
+ * @returns {CompanyService}
+ */
+export function useCompanyState() {
```
```suggestion
export function useCompanyService() {
```

aab-odoo commented on addons/web/static/src/webclient/company_service.js:
```diff
@@ -84,43 +64,142 @@ export const companyService = {
+export const companyService = {
+    dependencies: ["user", "router", "action"],
+    start(env, services) {
+        return new CompanyService(env, services);
```
I'm not sure about the class, because now internals of the service are accessible in the service API (even tough they're private and should not be used). I think functions in the scope better separates the private concerns and the public API. Also, it makes the public API clearer when you debug as you only see what you can really access. That's a personal opinion, we can discuss about it if you disagree.

BastienFafchamps replied:
Changed it to a function 👍

aab-odoo commented on addons/web/static/src/webclient/switch_company_menu/switch_company_menu.js:
```diff
@@ -2,65 +2,58 @@
-        this.companyService = useService("company");
-        this.store = useState(store);
+        this.companyService = useCompanyState();
+        this.debouncedApply = debounce(
```
this debounce logic of committing the state could be moved to the service as well

aab-odoo replied:
Actually, I think it would give more sense to the selectedCompanies attribute, as the debounce is (as far as I understand) the reason why we need it

aab-odoo commented:
@BastienFafchamps I'm still wondering whether the selected state must be in the service or not. It looks like a state of the menu instead, but I see it's used in account_reports. Why?

BastienFafchamps commented:
@aab-odoo I was actually wondering the same. I made another commit where I've extracted the "selectedCompanies" from the company service and it's now only used by the switch_company_menu. The company service is now way more simple and I've reverted it back to a function.

I've also renamed "availableCompanies" to "allowedCompanies" to stay coherent with the python side.

aab-odoo commented:
Hello @william-andre 

Any chance you or someone of your team could take a look at this PR any time soon? We would like to merge it before the freeze as well. Thanks :-)

william-andre reviewed (approved):
Seems to be working perfectly, tested with company branches as well

I don't really know why the change is needed since it just seems like refactoring to refactor but :shrug:

aab-odoo commented:
robodoo r+
